mobile_verifier/src/reward_share.rs

```diff
@@ -0,0 +1,335 @@
+    ) -> Result<(OwnerShares, MissingOwnerShares)> {
+        let mut owner_shares = OwnerShares::new();
+        let mut missing_owner_shares = MissingOwnerShares::new();
+        for (hotspot, share) in hotspot_shares {
+            if let Some(owner) = self.resolve_owner(&hotspot).await? {
+                *owner_shares.entry(owner).or_default() += share;
+            } else {
+                *missing_owner_shares.entry(hotspot).or_default() += share;
+            }
+        }
+        Ok((owner_shares, missing_owner_shares))
+    }
+}
+
+#[async_trait::async_trait]
+impl OwnerResolver for follower::Client<Channel> {
+    async fn resolve_owner(&mut self, address: &PublicKey) -> Result<Option<PublicKey>> {
+        let req = FollowerGatewayReqV1 {
+            address: address.to_vec(),
+        };
+        let res = self.find_gateway(req).await?.into_inner();
+
+        if let Ok(pub_key) = PublicKey::try_from(res.owner) {
+            return Ok(Some(pub_key));
+        }
+
+        Ok(None)
+    }
+}
+
+#[derive(Debug, Clone, Serialize)]
+pub struct GatheredShares {
+    pub shares: Shares,
+    pub invalid_shares: InvalidShares,
+    pub speed_shares: SpeedShares,
+    pub speed_shares_moving_avg: SpeedShareMovingAvgs,
+    pub invalid_speed_shares: InvalidSpeedShares,
+}
+
+impl GatheredShares {
+    pub async fn from_stream(
+        stream: &mut BytesMutStream,
+        after_utc: u64,
+        before_utc: u64,
+    ) -> Result<Self> {
+        let mut shares = Self {
+            shares: Shares::new(),
+            invalid_shares: InvalidShares::new(),
+            speed_shares: SpeedShares::new(),
+            speed_shares_moving_avg: SpeedShareMovingAvgs::default(),
+            invalid_speed_shares: InvalidSpeedShares::new(),
+        };
+
+        while let Some(Ok(msg)) = stream.next().await {
+            // NOTE: This will early exit with an error if we fail to decode
+
+            if let Ok(cell_heartbeat_req_v1) = CellHeartbeatReqV1::decode(msg.clone()) {
+                shares.gather_heartbeat(cell_heartbeat_req_v1, after_utc, before_utc)
+            } else if let Ok(speedtest_req_v1) = SpeedtestReqV1::decode(msg) {
+                shares.gather_speedtest(speedtest_req_v1, after_utc, before_utc)
+            } else {
+                continue;
+            }
+        }
+        Ok(shares)
+    }
+
+    fn gather_speedtest(
+        &mut self,
+        speedtest_req_v1: SpeedtestReqV1,
+        after_utc: u64,
+        before_utc: u64,
+    ) {
+        let SpeedtestReqV1 {
+            pub_key: st_pub_key,
+            timestamp: st_timestamp,
+            upload_speed: st_upload_speed,
+            download_speed: st_download_speed,
+            latency: st_latency,
+            ..
+        } = speedtest_req_v1;
+
+        if st_timestamp < after_utc || st_timestamp >= before_utc {
+            self.invalid_speed_shares.push(SpeedShare {
+                pub_key: PublicKey::try_from(st_pub_key.as_ref()).unwrap(),
```

Comment:
why not create `From<SpeedtestReqV1> for SpeedtestShare`? That not possible?

Reply:
Not really - we also need validity, so we'd have to have something like `fn from_speedtest_req_v1(req: SpeedtestReqV1, validity: Validity) -> Self`, which is probably cleaner

Reply:
you could have a From that makes validity false by default and then set it in above function you define there?

Reply:
I'm not a huge fan of that since it would lead to having code like: 
```rust
let mut share = SpeedShare::from(share);
share.validity = Validity::Invalid;
```
I'd like to come back and address this with a better, functional pattern.

Reply:
I'll address this in the `FileStore` refactor I'm working on

Reply:
Nevermind. Addressed